clients/drcachesim/drpt2trace/pt2ir.cpp

```diff
@@ -255,8 +296,111 @@ pt2ir_t::init(IN pt2ir_config_t &pt2ir_config, IN int verbosity)
+    uint64_t pt_pkt_decoder_offset = 0;
+    uint64_t pt_pkt_decoder_sync_offset = 0;
+
+    /* Backup the offset and sync offset of the decoder. */
+    int errcode = pt_pkt_get_offset(pt_pkt_decoder, &pt_pkt_decoder_offset);
+    if (errcode < 0) {
+        VPRINT(0, "Failed to get packet decoder offset: %s.\n",
+               pt_errstr(pt_errcode(errcode)));
+        return false;
+    }
+    errcode = pt_pkt_get_sync_offset(pt_pkt_decoder, &pt_pkt_decoder_sync_offset);
+    if (errcode < 0) {
+        VPRINT(0, "Failed to get packet decoder sync offset: %s.\n",
+               pt_errstr(pt_errcode(errcode)));
+        return false;
+    }
+
+    /* Attempt to perform pre-decoding to:
+     * (1) Determine the offset of the final data packet, which will serve as the stop
+     * position.
+     * (2) Locate the last Packet Stream Boundary (PSB) packet within the current buffer.
+     * Once identified, record its offset value. This offset value will then be used as
+     * the reference point for the subsequent left shift operation on the buffer.
+     */
+    pt_decoder_stop_offset_ = 0;
+    pt_raw_buffer_lshift_offset_ = 0;
+    uint64_t offset = pt_pkt_decoder_offset;
+    for (;;) {
+        struct pt_packet packet;
+        errcode = pt_pkt_next(pt_pkt_decoder, &packet, sizeof(packet));
+        if (errcode < 0) {
+            VPRINT(0, "Failed to get next packet: %s.\n", pt_errstr(pt_errcode(errcode)));
+            return false;
+        }
+        if (packet.type > ppt_pad) {
+            pt_decoder_stop_offset_ = offset + packet.size;
+        }
+        if (packet.type == ppt_psb) {
+            pt_raw_buffer_lshift_offset_ = offset;
+        }
+
+        errcode = pt_pkt_get_offset(pt_pkt_decoder, &offset);
+        if (errcode < 0) {
+            VPRINT(0, "Failed to get offset of packet decoder: %s.\n",
+                   pt_errstr(pt_errcode(errcode)));
+            return false;
+        }
+        if (offset >= pt_raw_buffer_data_size_) {
+            break;
+        }
+    }
+    /* Reset the packet decoder to the beginning of the trace. */
+    pt_pkt_decoder_set_pos(pt_pkt_decoder, pt_raw_buffer_.get() + pt_pkt_decoder_offset);
+    pt_pkt_decoder_set_sync(pt_pkt_decoder,
+                            pt_raw_buffer_.get() + pt_pkt_decoder_sync_offset);
+
+    return true;
+}
+
+bool
+pt2ir_t::post_decode()
+{
+    ASSERT(pt_instr_decoder_ != nullptr, "");
+    ASSERT(pt_raw_buffer_data_size_ > pt_raw_buffer_lshift_offset_, "");
+
+    if (pt_raw_buffer_lshift_offset_ != 0) {
+        struct pt_packet_decoder *pt_pkt_decoder =
+            pt_instr_decoder_get_pkt_decoder(pt_instr_decoder_);
+
+        /* Implement a left-shift operation on the buffer to enable continuous stream
+         * decoding.
+         */
+        pt_raw_buffer_data_size_ -= pt_raw_buffer_lshift_offset_;
+        memcpy(pt_raw_buffer_.get(), pt_raw_buffer_.get() + pt_raw_buffer_lshift_offset_,
```

Comment:
If there's overlap in the source and dest buffer ranges, would this continue to work as expected?